pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java

```diff
@@ -63,99 +65,147 @@ public PinotTaskManager(PinotHelixTaskResourceManager helixTaskResourceManager,
+   * Public API to schedule tasks (all task types) for all tables. It might be called from the non-leader controller.
+   * Returns a map from the task type to the task scheduled.
    */
   public synchronized Map<String, String> scheduleTasks() {
-    Map<String, String> tasksScheduled = scheduleTasks(_pinotHelixResourceManager.getAllTables());
-
-    // Reset the task because this method will be called from the Rest API instead of the periodic task scheduler
-    // TODO: Clean up only the non-leader tables instead of all tables
-    cleanUpTask();
-    setUpTask();
-
-    return tasksScheduled;
+    return scheduleTasks(_pinotHelixResourceManager.getAllTables(), false);
   }
 
   /**
-   * Check the Pinot cluster status and schedule new tasks for the given tables.
-   *
-   * @param tableNamesWithType List of table names with type suffix
-   * @return Map from task type to task scheduled
+   * Helper method to schedule tasks (all task types) for the given tables that have the tasks enabled. Returns a map
+   * from the task type to the task scheduled.
    */
-  private synchronized Map<String, String> scheduleTasks(List<String> tableNamesWithType) {
+  private synchronized Map<String, String> scheduleTasks(List<String> tableNamesWithType, boolean isLeader) {
     _controllerMetrics.addMeteredGlobalValue(ControllerMeter.NUMBER_TIMES_SCHEDULE_TASKS_CALLED, 1L);
 
     Set<String> taskTypes = _taskGeneratorRegistry.getAllTaskTypes();
-    int numTaskTypes = taskTypes.size();
-    Map<String, List<TableConfig>> enabledTableConfigMap = new HashMap<>(numTaskTypes);
-
-    for (String taskType : taskTypes) {
-      enabledTableConfigMap.put(taskType, new ArrayList<>());
-
-      // Ensure all task queues exist
-      _helixTaskResourceManager.ensureTaskQueueExists(taskType);
-    }
+    Map<String, List<TableConfig>> enabledTableConfigMap = new HashMap<>();
 
     // Scan all table configs to get the tables with tasks enabled
     for (String tableNameWithType : tableNamesWithType) {
       TableConfig tableConfig = _pinotHelixResourceManager.getTableConfig(tableNameWithType);
-      if (tableConfig != null) {
-        TableTaskConfig taskConfig = tableConfig.getTaskConfig();
-        if (taskConfig != null) {
-          for (String taskType : taskTypes) {
-            if (taskConfig.isTaskTypeEnabled(taskType)) {
-              enabledTableConfigMap.get(taskType).add(tableConfig);
-            }
+      if (tableConfig != null && tableConfig.getTaskConfig() != null) {
+        Set<String> enabledTaskTypes = tableConfig.getTaskConfig().getTaskTypeConfigsMap().keySet();
+        for (String enabledTaskType : enabledTaskTypes) {
+          if (taskTypes.contains(enabledTaskType)) {
+            enabledTableConfigMap.computeIfAbsent(enabledTaskType, k -> new ArrayList<>()).add(tableConfig);
+          } else {
+            LOGGER.warn("Task type: {} is not registered, cannot enable it for table: {}", enabledTaskType,
```

Comment:
Is it possible that we return this info to the client-side? I feel this information is useful for users to test and debug.

Reply:
Good point. We cannot directly return the message without backward-incompatible change on the rest endpoint (currently returns the map from task type to task scheduled). I made the change so that the returned map contains all the task types within the table configs, and if the task type is not registered, there will be no task scheduled (value is `null` for the task type), and then user can look into the log and find the reason.

Reply:
true,  maybe we add a new array field like "errors" later to hold these information.